src/NuGet.Jobs.RegistrationComparer/RegistrationComparerCollectorLogic.cs

```diff
@@ -0,0 +1,114 @@
+            ILogger<RegistrationComparerCollectorLogic> logger)
+        {
+            _utility = utility ?? throw new ArgumentNullException(nameof(utility));
+            _comparer = comparer ?? throw new ArgumentNullException(nameof(comparer));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public Task<IEnumerable<CatalogCommitItemBatch>> CreateBatchesAsync(IEnumerable<CatalogCommitItem> catalogItems)
+        {
+            return Task.FromResult(_utility.CreateSingleBatch(catalogItems));
+        }
+
+        public async Task OnProcessBatchAsync(IEnumerable<CatalogCommitItem> items)
+        {
+            var packageIdGroups = items
+                .GroupBy(x => x.PackageIdentity.Id, StringComparer.OrdinalIgnoreCase)
+                .Select(g => new
+                {
+                    Id = g.Key.ToLowerInvariant(),
+                    Versions = g
+                        .Select(x => x.PackageIdentity.Version)
+                        .Distinct()
+                        .OrderBy(x => x)
+                        .Select(x => x.ToNormalizedString().ToLowerInvariant())
+                        .ToList(),
+                })
+                .ToList();
+
+            _logger.LogInformation("Comparing {Count} package IDs.", packageIdGroups.Count);
+
+            var hiveGroups = _options
+                .Value
+                .Registrations
+                .SelectMany(x => new[]
+                {
+                    new { Hive = nameof(x.LegacyBaseUrl), Url = x.LegacyBaseUrl.TrimEnd('/') + '/' },
+                    new { Hive = nameof(x.GzippedBaseUrl), Url = x.GzippedBaseUrl.TrimEnd('/') + '/' },
+                    new { Hive = nameof(x.SemVer2BaseUrl), Url = x.SemVer2BaseUrl.TrimEnd('/') + '/' },
+                })
+                .GroupBy(x => x.Hive, x => x.Url);
+
+            var allWork = new ConcurrentBag<Func<Task>>();
+            var failures = new ConcurrentBag<string>();
+            foreach (var group in packageIdGroups)
+            {
+                foreach (var hiveGroup in hiveGroups)
+                {
+                    var baseUrls = hiveGroup.ToList();
+                    var hive = hiveGroup.Key;
+                    var id = group.Id;
+                    var versions = group.Versions;
+                    allWork.Add(async () =>
+                    {
+                        _logger.LogInformation("Verifying hive {Hive} for {PackageId}.", hive, id);
+                        try
+                        {
+                            await _comparer.CompareAsync(baseUrls, id, versions);
+                        }
+                        catch (Exception ex)
+                        {
+                            failures.Add(ex.Message);
+                            ResultWriter.WriteError(ex.Message);
```

Comment:
The real exception may be in an inner exception. Maybe use `.ToString()` instead?

Reply:
Sure, sounds fine.